cpp/include/cudf/detail/copy_range.cuh

```diff
@@ -0,0 +1,308 @@
+ * @param stream CUDA stream to run this function
+ */
+template <typename SourceValueIterator, typename SourceValidityIterator>
+void copy_range(SourceValueIterator source_value_begin,
+                SourceValidityIterator source_validity_begin,
+                mutable_column_view& target,
+                size_type target_begin, size_type target_end,
+                cudaStream_t stream = 0) {
+  CUDF_EXPECTS((target_begin <= target_end) &&
+                 (target_begin >= 0) &&
+                 (target_begin < target.size()) &&
+                 (target_end <= target.size()),
+               "Range is out of bounds.");
+  using T = typename std::iterator_traits<SourceValueIterator>::value_type;
+
+  // this code assumes that source and target have the same type.
+  CUDF_EXPECTS(type_to_id<T>() == target.type().id(), "the data type mismatch");
+
+  static_assert(warp_size == cudf::util::size_in_bits<bitmask_type>(),
+    "copy_range_kernel assumes bitmask element size in bits == warp size");
+
+  T * __restrict__ data = target.head<T>();
+  bitmask_type * __restrict__ bitmask = target.null_mask();
+  auto offset = target.offset();
+
+#if 1
+  auto warp_aligned_begin_lower_bound =
+    size_type{target_begin - (target_begin % warp_size)};
+  auto warp_aligned_end_upper_bound =
+    size_type{
+      (target_end % warp_size) == 0 ? \
+        target_end : target_end + (warp_size - (target_end % warp_size))
+    };
+  auto num_items =
+    warp_aligned_end_upper_bound - warp_aligned_begin_lower_bound;
+#else
+  // This one results in a compiler internal error! TODO: file NVIDIA bug
+  // size_type num_items =
+  //   cudf::util::round_up_safe(target_end - target_begin, warp_size);
+  // number threads to cover range, rounded to nearest warp
+  // this code runs for one additional round if target_begin is not warp
+  // aligned, and target_end is block_size + 1
+  auto num_items =
+    size_type{
+      warp_size *
+        cudf::util::div_rounding_up_safe(target_end - target_begin, warp_size)
+    };
+#endif
+
+  constexpr auto block_size = int{256};
+  static_assert(block_size <= 1024,
+    "copy_range kernel assumes block_size is not larger than 1024");
+
+  auto grid = cudf::util::cuda::grid_config_1d{num_items, block_size, 1};
+
+  if (target.nullable() == true) {
+    // TODO: if null_count is UNKNOWN_NULL_COUNT, no need to update null
+    // count (if null_count is UNKNOWN_NULL_COUNT, invoking null_count()
+    // will scan the entire bitmask array, and this can be surprising
+    // in performance if the copy range is small and the column size is
+    // large).
+    rmm::device_scalar<size_type> null_count(target.null_count(), stream);
+
+    auto kernel =
+      copy_range_kernel<SourceValueIterator, SourceValidityIterator, T, true>;
+    kernel<<<grid.num_blocks, block_size, 0, stream>>>(
+      source_value_begin, source_validity_begin,
+      data, bitmask, null_count.data(), offset + target_begin, offset + target_end);
```

Comment:
I am using _offset_ to update null_mask in cudf::bitmask_type granularity inside _copy_range_kernel_.
```
cudf::bitmask_type old_mask = bitmask[mask_idx];
if (lane_id == leader_lane) {
  new_mask = ...;
  bitmask[mask_idx] = new_mask;
  ...
}
```
Without _offset_ and just seeing _target_begin_ and _target_end_, I cannot figure out where the boundaries are.

If this code is bothering, I can pass mutable_column_device_view, target_begin, and target_end instead of data (target.head()), bitmask (target.null_mask()), offset + target_begin, and offset + target_end, but then I need to access offset inside _copy_range_kernel_.

Let me know what you think or you have a better way to handle this.

Reply:
Yes, I think it would be better to pass the output as a `mutable_column_device_view`.